Translate Lists and Keys page to French (#15)
reactjs/fr.reactjs.org: content/docs/lists-and-keys.md

next: forms.html
---

Tout d'abord, voyons comment transformer des listes en JavaScript.

Avec le code suivant, on utilise la méthode [`map()`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/map) pour prendre un tableau de nombres et doubler leurs valeurs. On assigne le nouveau tableau retourné par `map()` à une variable `doubled` et on l'affiche à la console :

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

Ce code affiche `[2, 4, 6, 8, 10]` à la console.

Avec React, transformer un tableau en une liste d'[éléments](/docs/rendering-elements.html) est presque identique.

### Afficher Plusieurs Composants {#rendering-multiple-components}

On peut construire des collections d'éléments et [les inclure dans du JSX](/docs/introducing-jsx.html#embedding-expressions-in-jsx) en utilisant les accolades `{}`.

Ci-dessous, on itère sur le tableau de nombres en utilisant la méthode JavaScript [`map()`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/map). On retourne un élément `<li>` pour chaque entrée du tableau.

> Enfin, on assigne le tableau d'éléments résultant à `listItems` :

Author: Non aucune raison. Probablement l'habitude (je saute beaucoup de lignes quand j'écris des documents)

);
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)

## Les clés {#keys}

> Les clés aident React à identifier quels éléments d'une liste ont changé, ont été ajoutés ou supprimés. Vous devez donner une clé à chaque élément dans un tableau afin de donner aux éléments une identité stable :

Reviewer: ```suggestion
Les clés aident React à identifier quels éléments d'une liste ont changé, ont été ajoutés ou supprimés. Vous devez donner une clé à chaque élément dans un tableau afin d’apporter aux éléments une identité stable :
```

C'est globalement nickel, j'aimerais juste retirer la répétition proche de « donner »

);
```

> Le meilleur moyen de choisir une clé est d'utiliser une chaîne de caractères qui identifie de façon unique un élément d'une liste parmi ses voisins. Le plus souvent on utilise l'ID de notre donnée comme clé :

Reviewer: ```suggestion
Le meilleur moyen de choisir une clé est d'utiliser une donnée qui identifie de façon unique un élément d'une liste parmi ses voisins. Le plus souvent on utilise l'ID de notre donnée comme clé :
```

Ta traduction est fidèle (👏) mais j'aimerais éviter de propager en VF cette restriction incorrecte aux *strings* : une clé peut être de n’importe quel type, en fait.

Author: Bon à savoir ! Je me demandais justement. Est-ce que ça vaut le coup d'enlever tous les `number.toString()` dans le code du coup ? (Je peux aussi en faire une PR sur la VO).

Reviewer: Franchement je préfère éviter qu'on touche à la logique du code dans les trads ; je préfère qu'on fasse remonter nos remarques à Dan via le dépôt canonique, et si ça bouge en amont, ça repercolera chez nous ensuite.

);
```

> Quand vous n'avez pas d'ID stable pour les éléments rendus, vous pouvez utiliser l'index de l'élément en dernier recours :

Reviewer: ```suggestion
Quand vous n'avez pas d'ID stable pour les éléments affichés, vous pouvez utiliser l'index de l'élément en dernier recours :
```

Toujours chercher à éviter *rendre* comme verbe…

```js{2,3}
const todoItems = todos.map((todo, index) =>

>   // Faites uniquement ceci si les éléments n'ont pas d'ID stable

Reviewer: ```suggestion
  // Ne faites ceci que si les éléments n'ont pas d'ID stable
```

C'est vraiment super déconseillé, du coup j'opterais pour une formulation encore plus négative, comme ceci.

  <li key={index}>
    {todo.text}
  </li>
);
```

> Nous ne recommandons pas d'utiliser l'index comme clé si l'ordre des éléments change. Cela peut avoir un effet négatif sur les performances, et peut causer des problèmes avec l'état du composant. Vous pouvez lire l'article de Robin Pokorny pour une [explication en profondeur de l'impact négatif de l'utilisation de l'index comme clé](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318) (en anglais). Si vous choisissez de ne pas donner explicitement de clé aux éléments d'une liste, React utilisera les index par défaut.

Reviewer: ```suggestion
Nous vous recommandons de ne pas utiliser l'index comme clé si l'ordre des éléments est susceptible de changer. Ça peut avoir un effet négatif sur les performances, et causer des problèmes avec l'état du composant. Vous pouvez lire l'article de Robin Pokorny pour une [explication en profondeur de l'impact négatif de l'utilisation de l'index comme clé](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318) (en anglais). Si vous choisissez de ne pas donner explicitement de clé aux éléments d'une liste, React utilisera l’index par défaut.
```

- L’anglais *don’t recommend* est un euphémisme courant pour *recommend not to* : ce n’est pas tant qu’on ne *recommande pas* : on *recommande de ne pas*, ce qui est différent.
- Quelques retraits de répétitions superflues
- Bien d'avoir pensé à préciser la langue cible
- Tu as eu raison de garder *index* au singulier avant, on va rester là-dessus pour la dernière phrase.

> Voici une [explication en profondeur de la raison pour laquelle les clés sont nécessaires](/docs/reconciliation.html#recursing-on-children) si vous voulez en apprendre plus.

Reviewer: ```suggestion
Si vous voulez en apprendre davantage, consultez cette [explication en profondeur de la raison pour laquelle les clés sont nécessaires](/docs/reconciliation.html#recursing-on-children).
```

Contextualiser en amont la raison du lien est plus lisible, et « voici » indique plus un contenu qui suit immédiatement qu’un lien vers une autre page.

### Extraire des Composants Avec des Clés {#extracting-components-with-keys}

Les clés n'ont une signification que dans le contexte du tableau qui les entoure.

> Par exemple, si on [extrait](/docs/components-and-props.html#extracting-components) un composant `ListItem`, on doit garder la clé sur l'élément `<ListItem />` dans le tableau, et non pas sur l'élément `<li>` dans le composant `ListItem` lui-même.

Reviewer: ```suggestion
Par exemple, si on [extrait](/docs/components-and-props.html#extracting-components) un composant `ListItem`, on doit garder la clé sur l'élément `<ListItem />` dans le tableau, et non sur l'élément `<li>` dans le composant `ListItem` lui-même.
```

Nickel, sauf que « non pas » est vraiment moche et vieilli, on retire toujours le « pas »

> **Exemple : Utilisation des clés erroné**

Reviewer: ```suggestion
**Exemple : Utilisation erronée des clés**
```

- Un adjectif est plus lisible lorsqu'il est épithète ou attribut qu'en fin de locution
- Faute d'accord

);
```

**Exemple : Utilisation correcte des clés**

```javascript{2,3,9,10}
function ListItem(props) {
  // Correct ! Pas la peine de spécifier la clé ici :
  return <li>{props.value}</li>;
}

function NumberList(props) {
  const numbers = props.numbers;
  const listItems = numbers.map((number) =>
    // Correct ! La clé doit être spécifiée dans le tableau.

>     <ListItem key={number.toString()}

Reviewer: On va pas changer le code, mais ce `.toString()` n’a tellement rien à foutre là…

Author: PR sur la doc originale ?

Reviewer: C'est ça, mais avant je vais discuter le bout de gras avec Dan sur le sujet, pour bien border le truc.

);
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

> Une règle à garder en tête est que chaque élément à l'intérieur d'un appel à `map()` a besoin d'une clé.

Reviewer: ```suggestion
Gardez en tête cette règle simple : chaque élément à l'intérieur d'un appel à `map()` a besoin d'une clé.
```

Les formulations « *Locution longue* est que *locution longue* » sont ampoulées en français, on essaie de reformuler.

> ### Les Clés Doivent Uniquement Être Uniques Parmi Leurs Voisins {#keys-must-only-be-unique-among-siblings}

Reviewer: ```suggestion
### Les Clés n’ont Besoin d’Être Uniques qu’au Sein de la Liste {#keys-must-only-be-unique-among-siblings}
```

Ahem…

Author: Oui j'ai un peu galéré sur cette traduction 😅

}

const posts = [
  {id: 1, title: 'Bonjour, monde', content: 'Bienvenue sur la documentation de React !'},
  {id: 2, title: 'Installation', content: 'Vous pouvez installer React depuis npm.'}
];
ReactDOM.render(
  <Blog posts={posts} />,
  document.getElementById('root')
);
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)

> Les clés servent d'indicateur à React mais ne sont pas passées à vos composants. Si vous avez besoin de la même valeur à l'intérieur de votre composant, passez-le dans une prop avec un nom différent :

Reviewer: ```suggestion
Les clés servent d'indicateur à React mais ne sont pas passées à vos composants. Si vous avez besoin de la même valeur dans de votre composant, passez-la dans une prop avec un nom différent :
```

Inutile d'alourdir le « dans » parfaitement acceptable avec un « à l’intérieur » (on préfèrerait d’ailleurs « au sein »)

);
```

> Avec l'exemple ci-dessus, le composant `Post` peut accéder à `props.id`, mais pas à `props.key`.

Reviewer: ```suggestion
Dans l'exemple ci-dessus, le composant `Post` peut accéder à `props.id`, mais pas à `props.key`.
```

Attention à la traduction trop littérale

### Intégrer map() dans du JSX {#embedding-map-in-jsx}

> Dans l'exemple précédent, nous avons déclaré séparément la variable `listItems` et l'avons inclue dans le JSX :

Reviewer: ```suggestion
Dans les exemples précédents, nous déclarions séparément la variable `listItems` pour ensuite l’inclure dans le JSX :
```

Plusieurs exemples référencés dans la VO.

}
```

> JSX autorise [d'intégrer des expressions](/docs/introducing-jsx.html#embedding-expressions-in-jsx) entre accolades. Nous pouvons donc utiliser `map()` dans notre code JSX :

Reviewer: ```suggestion
JSX permet [d'intégrer des expressions quelconques](/docs/introducing-jsx.html#embedding-expressions-in-jsx) entre accolades. Nous pouvons donc utiliser `map()` directement dans notre code JSX :
```

On « n’autorise pas de » en français 😉 On « autorise à » et on « permet de ». Le choix entre les deux dépend, mais en français « autoriser » indique plus l’octroi d’un droit que la mise à disposition d’une possibilité technique. Ici, “allow” indique clairement que JSX rend quelque chose possible, pas qu’il nous en *donne le droit*, d'où la préférence pour « permet ».

}
```

[**Essayer sur CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

> Ça donne un code plus lisible, mais ce style peut être utilisé de façon excessive. Comme JavaScript, c'est vous qui décidez quand ça vaut le coup d'extraire l'expression dans une variable pour plus de lisibilité. Gardez en tête que si le corps de `map()` est trop grand, il pourait être temps d'[extraire un composant](/docs/components-and-props.html#extracting-components).

Reviewer: ```suggestion
Ça rend parfois le code plus lisible, mais il faut éviter d’en abuser. Comme avec JavaScript, c'est vous qui décidez quand ça vaut le coup d'extraire l'expression dans une variable pour plus de lisibilité. Gardez en tête que si le corps de `map()` est trop profond ou trop riche, c’est sans doute le signe qu’il faudrait [extraire un composant](/docs/components-and-props.html#extracting-components).
```

Première phrase un peu trop pataude. Pour le reste, pas mal, mais il fallait affiner les caractéristiques visées du `map()` (la présence trop en profondeur / la taille en raison d'un code de rappel à niveaux multiples).
